hazelcast/src/main/java/com/hazelcast/dataconnection/DataConnectionBase.java

```diff
@@ -30,7 +30,7 @@
  * @since 5.3
  */
 @Beta
-public abstract class DataConnectionBase implements DataConnection {
+public abstract class DataConnectionBase implements DataConnection, AutoCloseable {
```

Comment:
If we introduce this, it should be added to `DataConnection` interface, but there we already have:

https://github.com/hazelcast/hazelcast/blob/64d68635edb819bf77b43ac6aec1640162bf876f/hazelcast/src/main/java/com/hazelcast/dataconnection/DataConnection.java#L148-L160

We would have `close` and `destroy` - we need to explain in the javadoc what is the difference.

Reply:
My vote is to apply `AutoCloseable`. `AutoCloseable` is a convenient API. I reviewed the code where we use retain/release and it doesn't feel unsafe to add it. Also, if somebody forgets to use `try-with-resources`, static analyzers either in IDE or on CI should remind about that miss.

Reply:
1. we call `retain` in many places, so it does not fit well into `AutoCloseable`, as `close` might not make the object destroyed.
2. IMO a better abstraction would be to return something `AutoCloseable` from `retain` and `getAndRetainDataConnection` (`DataConnectionInstance`?). For such object `close` would invoke `release` on underlying `DataConnectionBase`. 
3. There are not many places in _production code_ we we would use `DataConnection` in try-with-resources, often these are long lived. But even there, were we do, try-with-resources might be misleading, eg. replacing https://github.com/hazelcast/hazelcast/blob/master/extensions/kafka/src/main/java/com/hazelcast/jet/kafka/impl/StreamKafkaP.java#L399 with:

```java
        return (context) -> {
            try(KafkaDataConnection kafkaDataConnection = context
                    .dataConnectionService()
                    .getAndRetainDataConnection(dataConnectionRef.getName(), KafkaDataConnection.class)) {
                return kafkaDataConnection.newConsumer();
            }
        };
```
What happens to consumer the connection is closed? (I know the answer, but it is surprising for the reader)

Reply:
> We would have close and destroy - we need to explain in the javadoc what is the difference.

`close` = `release`, but this makes the API redundant

Reply:
I would keep retain/release because as @k-jamroz said, usually we keep dataconnection objects as long lived fields. As far I can see in this PR and briefly checking the code, we have only 2 places as candidates for `try-with-resources`

Reply:
@k-jamroz comments make sense.

Reply:
reverted the AutoCloseable on DataConnection : 76d67cac0cdca3689e50587fca7182bcc88cee41

Reply:
> From AutoCloseable#close JavaDoc. That means close does not have to be idempotent, can be called more than once.

yes, but also:

> However, implementers of this interface are strongly encouraged to make their close methods idempotent.

Reply:
> I have other problem with DataConnections - we create them, no clients, but DC is alive. We connect with them, release and the release kills DC

Check `DataConnectionBase.refCounter` - it is initially 1. The refCounter may reach 0 only when the DC is replaced with new version or deleted.